### ISSUE TITLE
Cannot properly mount exfat mircro sd card?

### ISSUE THREAD

**Issue by enriquezmark36** [diepquynh/android_device_samsung_scx35-common#3]
**TL;DR;** Can we specify `TARGET_EXFAT_DRIVER := exfat` since the kernel we use has the exfat driver?

--
So, I tried Android O and something has caught my eye. My micro sdcard, formatted as exfat, is mounted at /mnt/media_rw/ but it cannot be accessed like in Files app. You can through the root shell, though.
I got a lot of permission denied lines (due to mediaserver scanning files) from the sdcard process in the logcat.
What's weird is that the micro sdcard folder (e.g. /mnt/media_rw/E3B6-72A0) is owned by root and group root with something like a 600-ish permission. I couldn't remember the permission bits.
Yes, the link (e.g. /storage/E3B6-72A0) is not accessible.

Any tips other than what I wrote in the TL;DR; section ?

**Comment by diepquynh** [comment]
I didn't try exFAT so idk about your situation. I guess some kernel related stuff needed? 

Btw, we switch to sdfat in O instead of exFAT. sdfat is customized exFAT by Samsung

**Comment by enriquezmark36** [comment]
So do we have sdfat support right now I meant for the kernel side? AFAIK,
most Samsung kernels before Android N have an exfat fs driver.
BTW, I've meaning to ask, are there things not yet working in android O?

**Comment by diepquynh** [comment]
Yes, sdfat is in Samsung's O kernel (check universal9810 as an example)

There are tons of things that aren't working in O. For some devices (like core33g), camera doesn't work, as we're using hacks to make graphics working (which is frameworks_native patch btw). To make graphics work properly, for now from what I have investigate myself, we have to rewrite custom `gralloc_alloc_framebuffer_locked` that doesn't depend on `private_handle_t` since the default implementation has a member called `shared_fd`, which is always in negative number since framebuffer device doesn't use shared ION FD, but HIDL mapper treats it as the framebuffer fd so as a result, it couldn't clone the handle and we always have NULL cloned handle

**Comment by enriquezmark36** [comment]
So that's how it is. For the camera one, I thought it's just a missing
symbol that can be shim'd by libmemoryheapion or hex edited from the symbol
table, I think. Honestly, I am dumbfounded about gralloc. I have yet to it
study more.

Btw, I think asking about these through issues may not be the best, does
github has any pm functionality?

I'm gonna close this issue when I got to use the laptop. Nobody seems to
use exfat, except me, to experience this prob.

**Comment by enriquezmark36** [comment]
Excuse me for a bit @remilia15 , You may ignore this if you'd like.

About gralloc I started investigating it just now but I don't have the knowledge (to use the debug tools). I tried modified the code to print out every call to native_handle_clone(), I think I didn't have any issues with any cloned handle (on the contrary, all duplication and thus the clone creations have succeeded). 
I tagged the lines with `clone:`.

But what's more surprising is that I'm getting a lot of EGL_BAD_ALLOC and also gralloc2 allocate fails, even though alloc() from our gralloc succeeds. Here's the  [logcat](https://hastebin.com/oqegosepil.coffeescript) if you're interested.

I have yet to learn how to debug shared libraries and to find the full back trace including the stacks  before it commit errors.

You have reached the end of this message --- Thanks for reading --- .

**Comment by diepquynh** [comment]
Welp, our gralloc uses gralloc1 impl, not gralloc0 like you did. The passthrough implementation you used is incorrect so obviously it would return weird errors that you can't debug

**Comment by ngoquang2708** [comment]
My board is scx15, gralloc0 works with patched libui.

Edit: Sorry for spamming. I didn't noticed that you are talking about HIDL.

**Comment by diepquynh** [comment]
But hey, both sc8830 and scx15 are using gralloc1 as allocator though (since N I guess)

**Comment by ngoquang2708** [comment]
Yeah. It is graloc1 adaptor in libui in N. Sadly it is removed in O in favour of HIDL.

**Comment by diepquynh** [comment]
Well actually the adapter is moved to HIDL impl, but what's interfering is the mapper
